assets/src/edit-story/components/carousel/carouselProvider.js

```diff
@@ -0,0 +1,139 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import { useCallback, useLayoutEffect, useRef, useMemo, useState } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { useStory } from '../../app';
+import CarouselContext from './carouselContext';
+import useCarouselSizing from './useCarouselSizing';
+import useCarouselScroll from './useCarouselScroll';
+import useCarouselKeys from './useCarouselKeys';
+
+function CarouselProvider({ availableSpace, children }) {
+  const {
+    pages,
+    currentPageId,
+    setCurrentPage,
+    arrangePage,
+  } = useStory(
+    ({
+      state: { pages, currentPageId },
+      actions: { setCurrentPage, arrangePage },
+    }) => ({ pages, currentPageId, setCurrentPage, arrangePage })
+  );
+
+  const [listElement, setListElement] = useState(null);
+  const pageRefs = useRef([]);
+
+  const numPages = pages.length;
+
+  const pageIds = useMemo(() => pages.map(({ id }) => id), [pages]);
+
+  const {
+    pageThumbWidth,
+    pageThumbHeight,
+    pageThumbMargin,
+    carouselWidth,
+    hasOverflow,
+  } = useCarouselSizing({ availableSpace, numPages });
+
+  const setPageRef = useCallback((page, el) => {
+    pageRefs.current[page.id] = el;
+  }, []);
+
+  const {
+    canScrollBack,
+    canScrollForward,
+    scrollBack,
+    scrollForward,
+  } = useCarouselScroll({ listElement, carouselWidth, hasOverflow });
+
+  useCarouselKeys({ listElement, pageRefs });
+
+  useLayoutEffect(() => {
+    if (!hasOverflow) {
+      return;
+    }
+    const currentPageRef = pageRefs.current[currentPageId];
+
+    if (!currentPageRef || !currentPageRef.scrollIntoView) {
+      return;
+    }
+
+    currentPageRef.scrollIntoView({
+      inline: 'center',
+      behavior: 'smooth',
+    });
+  }, [currentPageId, hasOverflow, carouselWidth, pageRefs]);
+
+  const clickPage = useCallback((page) => setCurrentPage({ pageId: page.id }), [
+    setCurrentPage,
+  ]);
+
+  const rearrangePages = useCallback(
+    (oldPos, newPos) => {
+      const pageId = pageIds[oldPos];
+      arrangePage({ pageId, position: newPos });
+      setCurrentPage({ pageId });
+    },
+    [pageIds, arrangePage, setCurrentPage]
+  );
+
+  const value = {
+    state: {
+      pageThumbWidth,
+      pageThumbHeight,
+      pageThumbMargin,
+      availableSpace,
+      carouselWidth,
+      hasOverflow,
+      pages,
+      pageIds,
+      numPages,
+      currentPageId,
+      canScrollBack,
+      canScrollForward,
```

Comment:
Left a separate comment as well but if both `canScroll*`s are `false`, should the buttons be displayed at all?

Reply:
Hm, the design isn't clear on that. I'll ask, thanks!

Reply:
Forwarded to Sam

Reply:
We'll hide completely if no overflow. Fixed in d05008525.